tower-buffer/src/future.rs

```diff
@@ -35,36 +35,36 @@ where
     }
 }
 
-impl<T> Future for ResponseFuture<T>
+impl<T, A, B> Future for ResponseFuture<T>
 where
-    T: Future,
-    T::Error: Into<Error>,
+    T: Future<Output = Result<A, B>>,
+    B: Into<Error>,
 {
-    type Item = T::Item;
-    type Error = Error;
-
-    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
-        use self::ResponseState::*;
+    type Output = Result<A, Error>;
 
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
         loop {
             let fut;
 
             match self.state {
-                Failed(ref mut e) => {
-                    return Err(e.take().expect("polled after error"));
+                ResponseState::Failed(ref mut e) => {
+                    return Poll::Ready(Err(e.take().expect("polled after error")));
                 }
-                Rx(ref mut rx) => match rx.poll() {
-                    Ok(Async::Ready(Ok(f))) => fut = f,
-                    Ok(Async::Ready(Err(e))) => return Err(e.into()),
-                    Ok(Async::NotReady) => return Ok(Async::NotReady),
-                    Err(_) => return Err(Closed::new().into()),
+                ResponseState::Rx(ref mut rx) => match unsafe { Pin::new_unchecked(rx) }.poll(cx) {
```

Comment:
So, this `unsafe` is okay because the only time we move `rx` is when we re-assign `self.state` below, and that's the drop point for `rx`, and that's okay? Would be good to place that justification in a comment here.